module/zfs/vdev.c

```diff
@@ -898,6 +898,16 @@ vdev_alloc(spa_t *spa, vdev_t **vdp, nvlist_t *nv, vdev_t *parent, uint_t id,
 	(void) nvlist_lookup_uint64(nv, ZPOOL_CONFIG_CREATE_TXG,
 	    &vd->vdev_crtxg);
 
+	if (vd->vdev_ops == &vdev_root_ops &&
+	    (alloctype == VDEV_ALLOC_LOAD ||
+	    alloctype == VDEV_ALLOC_SPLIT ||
+	    alloctype == VDEV_ALLOC_ROOTPOOL)) {
+		(void) nvlist_lookup_uint64(nv,
+		    ZPOOL_CONFIG_VDEV_ROOT_ZAP, &spa->spa_root_vdev_zap);
+	} else {
+		ASSERT0(spa->spa_root_vdev_zap);
```

Comment:
What's the rational for storing this in the spa rather than the root vdev?  Modeling this after how the top zaps are stored seems preferable; i.e. `vd->vdev_root_zap`.  This is why we're running afoul of that new ASSERT0 since we're checking `spa->spa_root_vdev_zap` is zero for all vdev types.

http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/23347/steps/shell_4/logs/console

Reply:
The rationale was that it was less wasteful (in terms of space) to store the root vdev zap in spa_t as opposed to vdev_t since `vdev_root_zap` will be zero except for all vdevs except the root vdev.

With that said, my first approach did store the root vdev zap in `vd->vdev_root_zap` and tracked it in the AVZ. I think this would require a feature flag for the root vdev zap otherwise `vdev_count_verify_zaps()` will panic when importing a pool using an older module.

I agree with you that the root vdev zap fits better into the system by following the same pattern as top zaps.

Reply:
bump

Should I update this PR to have the root vdev zap follow the same pattern as top/leaf zaps then?

The one drawback here is that it would require adding a feature flag for adding the root vdev zap to the AVZ and not being backwards compatible with pools that don't have this feature enabled.

As I said earlier, I agree that the approach above fits into the current design more naturally - but doesn't allow for backwards compatibility with older pools.

Just want to double check with the direction to go before updating this pull request.

Reply:
Well that's unfortunate.  I see why you opted for this approach now.  In an ideal world I'd prefer this code be aligned with the approach taken for the top/leaf zaps.   However, we need retain backwards compatibility and if we can do so without introducing a feature flag so much the better.

Given that let's stick with your current approach.  If you can update the PR again to resolve the ASSERTs I think we can move forward with this.  Thanks for the reminder to look at this again.